store/postgres/src/store.rs

```diff
@@ -23,18 +23,204 @@ use graph::{tokio, tokio::timer::Interval};
+    None,
+}
+
+impl<T, U> EitherFuture<T, U>
+where
+    T: Future,
+    U: Future,
+{
+    fn new_left(child: T) -> EitherFuture<T, U> {
+        EitherFuture::Left(UnitFuture { child })
+    }
+
+    fn new_right(child: U) -> EitherFuture<T, U> {
+        EitherFuture::Right(UnitFuture { child })
+    }
+}
+
+impl<T, U> Future for EitherFuture<T, U>
+where
+    T: Future,
+    U: Future,
+{
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            EitherFuture::Left(uf) => uf.poll(),
+            EitherFuture::Right(uf) => uf.poll(),
+            EitherFuture::None => Ok(Async::Ready(())),
+        }
+    }
+}
+
+/// Internal representation of a Store subscription. Calling `subscribe`
+/// results in us storing an `EntityChange` sender, calling `block_ticks`
+/// results in us storing a `BlockTick` sender
+#[derive(Clone, Debug)]
+enum Subscription {
+    // We have entities in both kinds of subscriptions; since we clone the
+    // entire subscription in a few places, we make the entities an Arc
+    // to make the cloning cheaper.
+    Changes {
+        entities: Arc<Vec<SubgraphEntityPair>>,
+        sender: Sender<EntityChange>,
+    },
+    Ticks {
+        entities: Arc<Vec<SubgraphEntityPair>>,
+        sender: Sender<BlockTick>,
+    },
+}
+
+/// Track which subscriptions have had changes to their blocks. The entries
+/// in the set are the id's of subscriptions for which we've seen an
+/// EntityChange matching their interesting entities. When the corresponding
+/// BlockTick comes by, we pass a BlockTick to the receiver and remove
+/// the id from the set
+#[derive(Clone)]
+struct DirtyBlocks {
+    ids: Arc<RwLock<std::collections::HashSet<String>>>,
+}
+
+impl DirtyBlocks {
+    fn new() -> Self {
+        DirtyBlocks {
+            ids: Arc::new(RwLock::new(std::collections::HashSet::new())),
+        }
+    }
+
+    fn remove(&self, id: &str) -> bool {
+        self.ids.write().unwrap().remove(id)
+    }
+
+    fn insert(&self, id: String) -> bool {
+        self.ids.write().unwrap().insert(id)
+    }
+
+    fn contains(&self, id: &str) -> bool {
+        self.ids.read().unwrap().contains(id)
+    }
+}
+
+/// Subscription translates from the internal `StoreEvent` to either
+/// `EntityChange` or `BlockTick`, depending on what the subscriber is
+/// interested in. The 'other' kind of event for a subscriber is simply
+/// dropped on the floor
+///
+/// This is not a full implementation of a `futures::sink::Sink` since
+/// we don't need that right now, but could be expanded to that
+impl Subscription {
+    fn send(
+        self,
+        e: StoreEvent,
+        id: &String,
+        dirty_blocks: &DirtyBlocks,
+    ) -> impl Future<Item = (), Error = ()> {
+        match (self, e) {
+            (Subscription::Changes { sender, .. }, StoreEvent::Change(change)) => {
+                EitherFuture::new_left(sender.send(change))
```

Comment:
The custom `EitherFuture` doesn't seem very necessary. I'd just box the futures and return a boxed trait object, if you'd rather avoid that another option is to use a nested built-in [Either](https://docs.rs/futures/0.1.11/futures/future/enum.Either.html) so that you have three variants with `Either<T, Either<U, V>>`. The `UnitFuture` is just `.map(|_| ()).map_err(|_| ())` which doesn't seem to bad to write out.